### PR TITLE
Fix Pro FOUC reveal gating

### LINKED ISSUE #4031: Follow-up: Fix client-only component reveal until CSS and JS are ready
## Context

The RSC/streaming FOUC acceptance tests added in the pending test PR expose a real client-only reveal-gating bug.

For client-side-only components, the component should never appear until both its CSS and JS bundles have downloaded. The CSS-ready / JS-delayed path passed locally, but the JS-ready / CSS-delayed path did not.

## Failing command

```bash
cd react_on_rails_pro/spec/dummy &&
pnpm exec playwright test e2e-tests/rsc_fouc.spec.ts --project=chromium --reporter=line
```

## Failing test

- `client-only content does not appear until CSS loads, even when JS is ready`

## Observed behavior

- The generated client-only CSS file is linked in the document head.
- The Playwright route intercepts that CSS response by looking for `--client-only-fouc-probe-sentinel` in the CSS body and holds it.
- With JS available and the CSS response held, the client-only probe still becomes visible before the held CSS is released.

Error excerpt:

```text
Expected: not visible
Received: visible
Locator: getByTestId('client-only-fouc-probe')
```

## Evidence

Playwright error context from the local run:

- `react_on_rails_pro/spec/dummy/test-results/rsc_fouc-RSC-and-streaming-471d3-loads-even-when-JS-is-ready-chromium/error-context.md`

## Expected behavior matrix

| Scenario | Expected |
| --- | --- |
| Client-only CSS loaded, JS delayed | Probe is not visible until JS is released, then visible and styled |
| Client-only JS loaded, CSS delayed | Probe is not visible during a stable wait window |
| Client-only CSS released after JS-loaded / CSS-delayed scenario | Probe becomes visible and styled |

## Pending test reproduction

The pending test should be runnable with:

```bash
cd react_on_rails_pro/spec/dummy &&
REACT_ON_RAILS_RUN_PENDING_FOUC_TESTS=true pnpm exec playwright test e2e-tests/rsc_fouc.spec.ts --project=chromium --reporter=line
```

### LINKED ISSUE #4032: Follow-up: Fix RSC streamed component CSS reveal gating to prevent FOUC
## Context

The RSC/streaming FOUC acceptance tests added in the pending test PR expose a real streamed SSR reveal bug.

These tests intentionally assert user-visible behavior instead of implementation details:

- If the CSS required by a streamed RSC client boundary is loaded, the component should be visible and styled even if JS is delayed.
- If that CSS is not loaded yet, the streamed component should not become visible, regardless of whether JS has loaded.

## Failing command

```bash
cd react_on_rails_pro/spec/dummy &&
pnpm exec playwright test e2e-tests/rsc_fouc.spec.ts --project=chromium --reporter=line
```

## Failing tests

- `streamed RSC content is visible and styled when its CSS is loaded, even while JS is delayed`
- `streamed RSC content does not appear before its CSS when JS is available`
- `streamed RSC content still waits for CSS if JS finishes first`

## Observed behavior

- The streamed RSC HTML is inserted into the page while the component CSS is not applied.
- With JS delayed, the streamed RSC probe is visible but computed styles remain browser defaults:
  - `backgroundColor: rgba(0, 0, 0, 0)`
  - `borderTopColor: rgb(229, 231, 235)`
  - `color: rgb(0, 0, 0)`
  - CSS sentinel custom property is empty
- With CSS delayed and JS available, the RSC probe becomes visible before the held CSS response is released.
- With CSS delayed and JS delayed, releasing JS first still allows the RSC probe to become visible before the held CSS response is released.

The streamed response currently includes a style preload and then swaps the content into the DOM:

```html
<link rel="preload" as="style" href="/webpack/test/css/client1-46072b81.css"/>
```

That fetches the stylesheet but does not make it an applied stylesheet before the streamed content is revealed.

## Evidence

Playwright error contexts from the local run:

- `react_on_rails_pro/spec/dummy/test-results/rsc_fouc-RSC-and-streaming-308b7-ed-even-while-JS-is-delayed-chromium/error-context.md`
- `react_on_rails_pro/spec/dummy/test-results/rsc_fouc-RSC-and-streaming-ac4f4-ts-CSS-when-JS-is-available-chromium/error-context.md`
- `react_on_rails_pro/spec/dummy/test-results/rsc_fouc-RSC-and-streaming-10b38-or-CSS-if-JS-finishes-first-chromium/error-context.md`

## Expected behavior matrix

| Scenario | Expected |
| --- | --- |
| Required RSC CSS loaded, JS delayed | RSC probe is visible and styled |
| Required RSC CSS delayed, JS allowed | RSC probe is not visible during a stable wait window |
| Required RSC CSS delayed, JS delayed, then JS released first | RSC probe remains hidden until CSS is released |
| Required RSC CSS released after either delayed scenario | RSC probe becomes visible and styled |

## Pending test reproduction

The pending tests should be runnable with:

```bash
cd react_on_rails_pro/spec/dummy &&
REACT_ON_RAILS_RUN_PENDING_FOUC_TESTS=true pnpm exec playwright test e2e-tests/rsc_fouc.spec.ts --project=chromium --reporter=line
```

### PR DESCRIPTION
**Pull request by justin808**
Fixes #4031.
Fixes #4032.

## Summary
- Gate Pro client-only auto-loaded component mounting on generated stylesheet readiness, including shared generated CSS discovered from manifest href metadata.
- Promote streamed RSC client chunk stylesheet preloads to real stylesheet links before reveal, preserving preload attributes and handling split stream chunks.
- Enable and extend FOUC Playwright coverage for CSS-before-reveal, JS-before-CSS, client-only gating, and CSS chunk scope.
- Add the user-visible Pro fix to `CHANGELOG.md`.

## Validation
- `pnpm exec jest tests/ClientSideRenderer.test.ts tests/injectRSCPayload.test.ts --runInBand`
- `bundle exec rspec spec/helpers/react_on_rails_helper_spec.rb:63 spec/helpers/react_on_rails_helper_spec.rb:1402`
- `pnpm run build`
- `pnpm run type-check`
- `pnpm exec eslint packages/react-on-rails-pro/src/ClientSideRenderer.ts packages/react-on-rails-pro/src/injectRSCPayload.ts packages/react-on-rails-pro/tests/ClientSideRenderer.test.ts packages/react-on-rails-pro/tests/injectRSCPayload.test.ts react_on_rails_pro/spec/dummy/e2e-tests/rsc_fouc.spec.ts`
- `pnpm exec prettier --check packages/react-on-rails-pro/src/ClientSideRenderer.ts packages/react-on-rails-pro/src/injectRSCPayload.ts packages/react-on-rails-pro/tests/ClientSideRenderer.test.ts packages/react-on-rails-pro/tests/injectRSCPayload.test.ts react_on_rails_pro/spec/dummy/e2e-tests/rsc_fouc.spec.ts`
- `bundle exec rubocop react_on_rails/lib/react_on_rails/helper.rb react_on_rails/lib/react_on_rails/pro_helper.rb react_on_rails/spec/dummy/spec/helpers/react_on_rails_helper_spec.rb`
- `cd react_on_rails_pro/spec/dummy && pnpm run build:test`
- `cd react_on_rails_pro/spec/dummy && pnpm exec playwright test e2e-tests/rsc_fouc.spec.ts`
- `pnpm exec prettier --check CHANGELOG.md`
- `git diff --check`
- `/Users/justin/.agents/skills/autoreview/scripts/autoreview --mode branch --base origin/main`
- Pre-push hook: branch RuboCop and online markdown links passed; lychee reported 14 redirects and 0 errors.

## Review Notes
- Codex autoreview: clean, no accepted/actionable findings.
- Claude `/code-review origin/main` and `/simplify origin/main` were attempted with `claude-opus-4-8 --max-budget-usd 20`, but both CLI runs hung without output and returned only `Execution error` after interrupt.

---

> [!NOTE]
> **Medium Risk**
> Touches timing-sensitive Pro streaming and client hydration paths with substantial HTML stream rewriting; mitigated by scoped matching rules, inference timeouts, and fail-open behavior when CSS never loads.
> 
> **Overview**
> Fixes **Pro FOUC** where client-only roots and streamed RSC Suspense reveals could show unstyled content because hydration/reveal ran before generated or chunk CSS was applied.
> 
> **Client-only:** Pro now waits in parallel for the component bundle and for matching generated stylesheet `link` elements (from `data-generated-stylesheet-hrefs` on the spec tag plus `/generated/{Component}` href heuristics), with a 10s fail-open timeout. Rails emits that metadata when `auto_load_bundle` resolves manifest preload hrefs for `generated/{component}`.
> 
> **Streamed RSC:** `injectRSCPayload` promotes only RSC client-chunk `rel=preload as=style` links to real stylesheets (`data-precedence="rsc-css"`), infers extra stylesheet tags from Flight + `loadable-stats.json`, and can defer `$RC` reveal HTML until inference finishes—while still streaming fallbacks and handling link tags split across chunks or UTF-8 boundaries. App-authored style preloads are left unchanged.
> 
> Playwright FOUC acceptance tests for #4031/#4032 are enabled (no longer `fixme`); Jest coverage expanded for both paths.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 69251bdd4ae4efbf04e1457e128abf02eb994270. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

## Summary by CodeRabbit

* **Bug Fixes**
  * Fixed Flash of Unstyled Content (FOUC) in Pro with React Server Components by delaying client-only hydration until the component’s generated stylesheet links are ready (10s timeout) with “fails open” fallback.
  * Improved streamed RSC CSS handling by promoting streamed style preloads into real stylesheet links, handling links split across stream boundaries, and preserving app-authored preloads while ensuring correct reveal ordering.
* **Tests**
  * Expanded Jest and stream-related coverage for stylesheet gating, reveal deferral, and UTF-8 boundary streaming edge cases.
  * Updated Playwright RSC FOUC acceptance tests to run normally and improved timing/cleanup.

### CODE REVIEW

**Comment by justin808** [comment]
Follow-up from the failed `dummy-app-rspack-rsc-runtime-gate` run:

Root cause: the Rspack RSC stream can reveal client-component HTML from the Flight `$RC(...)` script before the corresponding client CSS chunk is discoverable as a streamed `<link rel="preload" as="style">`. Webpack exercised the existing preload-promotion path, but Rspack only exposed the client chunk through the Flight reference, while `loadable-stats.json` had the CSS mapping.

Fix in `704ed821f`: infer RSC client stylesheets from `loadable-stats.json` for `clientN` Flight references and flush those stylesheet links before streamed reveal HTML. I also added a regression test for that ordering and cleaned up Playwright routes after each FOUC test to prevent route callbacks from outliving the page/context.

Local validation:
- `pnpm exec jest tests/injectRSCPayload.test.ts --runInBand`
- `pnpm run build`
- `pnpm run type-check`
- `cd react_on_rails_pro/spec/dummy && bundle exec rake react_on_rails:generate_packs && pnpm run build:test:rspack`
- `cd react_on_rails_pro/spec/dummy && CI=true SHAKAPACKER_ASSETS_BUNDLER=rspack pnpm exec playwright test e2e-tests/rsc_echo_props.spec.ts e2e-tests/rsc_route_ssr_false.spec.ts e2e-tests/rsc_fouc.spec.ts --project=chromium`
- `cd react_on_rails_pro/spec/dummy && SHAKAPACKER_ASSETS_BUNDLER=rspack pnpm e2e-test:rsc`
- `pnpm run lint`
- `pnpm start format.listDifferent`
- `git diff --check`

File: packages/react-on-rails-pro/src/injectRSCPayload.ts [hunk 1]
```diff
@@ -90,6 +90,100 @@ function createRSCDiagnosticScript(
   );
 }
 
+const RSC_CLIENT_CHUNK_STYLESHEET_PATH = /\/css\/client\d+-[^/]+\.css$/;
+
+function getQuotedAttribute(tag: string, attributeName: string) {
+  const attributeMatch = tag.match(new RegExp(`\\s${attributeName}=(["'])(.*?)\\1`, 'i'));
+  return attributeMatch?.[2];
```
[the remark this answers is not in the document]
Comment (justin808) [reply packages/react-on-rails-pro/src/injectRSCPayload.ts:118]: Fixed in 915139ab5. `getQuotedAttribute` now escapes the attribute name before interpolating it into the `RegExp` constructor.

File: packages/react-on-rails-pro/src/ClientSideRenderer.ts [hunk 2]
```diff
@@ -44,6 +44,100 @@ import * as StoreRegistry from './StoreRegistry.ts';
...
+  const stylesheetLinks = Array.from(
+    document.querySelectorAll<HTMLLinkElement>('link[rel~="stylesheet"][href]'),
+  ).filter((link) => generatedStylesheetMatchesComponent(link, componentName, generatedStylesheetHrefs));
+
+  if (stylesheetLinks.length === 0) {
+    return Promise.resolve();
+  }
+
+  return Promise.all(stylesheetLinks.map(waitForStylesheet)).then(() => undefined);
+}
```
[the remark this answers is not in the document]
Comment (justin808) [reply packages/react-on-rails-pro/src/ClientSideRenderer.ts:141]: Handled in 915139ab5. Added the local assumption comment: generated stylesheet links are emitted into `<head>` before the component script runs.

File: packages/react-on-rails-pro/src/injectRSCPayload.ts [hunk 3]
```diff
@@ -213,14 +416,22 @@ export default function injectRSCPayload(
...
       offset += buffer.length;
     }
 
-    // 3. RSC payload chunk scripts LAST
+    // 3. HTML chunks THIRD
+    // Component markup that references the initialized arrays
+    if (gatedHtmlBuffer.length > 0) {
+      gatedHtmlBuffer.copy(combinedBuffer, offset);
+      offset += gatedHtmlBuffer.length;
+    }
```
[the remark this answers is not in the document]
Comment (justin808) [reply packages/react-on-rails-pro/src/injectRSCPayload.ts:543]: Fixed in 915139ab5. `flush()` now holds reveal-script HTML until initial RSC client stylesheet inference has either emitted the stylesheet link from Flight/loadable-stats data or the RSC stream has ended. Added a regression test where React calls `flush()` before the delayed Flight chunk arrives; it fails before the fix and now asserts the stylesheet link precedes the reveal HTML.

File: packages/react-on-rails-pro/src/ClientSideRenderer.ts [hunk 4]
```diff
@@ -196,7 +290,10 @@ class ComponentRenderer {
       const domNode = document.getElementById(domNodeId);
       if (domNode) {
         this.domNode = domNode;
-        const componentObj = await ComponentRegistry.getOrWaitForComponent(name);
+        const [componentObj] = await Promise.all([
+          ComponentRegistry.getOrWaitForComponent(name),
+          waitForGeneratedComponentStylesheets(name, el),
+        ]);
```
[the remark this answers is not in the document]
Comment (justin808) [reply packages/react-on-rails-pro/src/ClientSideRenderer.ts:297]: Current generated-auto-load flow emits the stylesheet `<link>` elements into `<head>` before the component script runs, so this code intentionally snapshots existing links rather than waiting for future dynamic insertion. 915139ab5 documents that ordering assumption at the query site; supporting late dynamic stylesheet injection would be a separate behavior change outside this FOUC fix.

File: packages/react-on-rails-pro/src/injectRSCPayload.ts [hunk 5]
```diff
@@ -213,14 +416,22 @@ export default function injectRSCPayload(
...
+    for (const buffer of rscClientStylesheetBuffers) {
       buffer.copy(combinedBuffer, offset);
       offset += buffer.length;
     }
 
-    // 3. RSC payload chunk scripts LAST
+    // 3. HTML chunks THIRD
+    // Component markup that references the initialized arrays
+    if (gatedHtmlBuffer.length > 0) {
+      gatedHtmlBuffer.copy(combinedBuffer, offset);
```
[the remark this answers is not in the document]
Comment (justin808) [reply packages/react-on-rails-pro/src/injectRSCPayload.ts]: Fixed in 915139ab5. Added a failing regression for the synchronous HTML `flush()` / delayed Flight timing, then changed the stream gate so reveal HTML waits for initial inferred RSC stylesheet readiness before it can flush.

File: packages/react-on-rails-pro/src/injectRSCPayload.ts [hunk 6]
```diff
@@ -380,6 +621,8 @@ export default function injectRSCPayload(
               const chunkBuf = chunk instanceof Uint8Array ? chunk : new TextEncoder().encode(chunk);
               parser.feed(chunkBuf, handleParsedChunk);
             }
+            hasResolvedInitialRSCClientStylesheetInference = true;
```
[the remark this answers is not in the document]
Comment (justin808) [reply packages/react-on-rails-pro/src/injectRSCPayload.ts]: Fixed in 70bf6dcf1. The reveal gate now tracks stylesheet inference per RSC payload stream instead of a request-wide boolean: each stream holds a pending inference slot until its first parsed Flight chunk or empty end, with a short fail-open timeout for long-lived Redis-style streams. Added regressions for multi-stream ordering and long-lived streams, and the Pro dummy e2e suite passed locally.

File: packages/react-on-rails-pro/src/injectRSCPayload.ts [hunk 7]
```diff
@@ -90,6 +92,247 @@ function createRSCDiagnosticScript(
...
+    }
+
+    if (hiddenBoundaryStartIndex !== -1) return hiddenBoundaryStartIndex;
+  }
+
+  return scriptStartIndex === -1 ? revealCallIndex : scriptStartIndex;
+}
+
+function splitReactSuspenseRevealHtmlBuffer(htmlBuffer: Buffer) {
+  const htmlString = htmlBuffer.toString('utf8');
```
[the remark this answers is not in the document]
Comment (justin808) [reply packages/react-on-rails-pro/src/injectRSCPayload.ts:225]: Fixed in 9ef364688. `splitReactSuspenseRevealHtmlBuffer` now uses the decoded HTML only to find the ASCII reveal split point, then converts that character index back to a byte offset and slices the original Buffer with `subarray()`. Added a regression that defers reveal HTML while `é` is split across HTML chunks, asserting the final stream preserves `café` without U+FFFD replacement characters.

File: packages/react-on-rails-pro/src/injectRSCPayload.ts [hunk 8]
```diff
@@ -90,6 +92,249 @@ function createRSCDiagnosticScript(
...
+    ? splitIncompleteLinkTagTail(htmlString)
+    : { completeHtml: htmlString, incompleteLinkTagTail: '' };
+  const gatedHtml = completeHtml.replace(
+    /<link\b(?=[^>]*\brel=(["'])(?:(?!\1).)*\bpreload\b(?:(?!\1).)*\1)(?=[^>]*\bas=(["'])style\2)(?=[^>]*\bhref=(["'])(?:(?!\3).)+\3)[^>]*\/?>/gi,
+    (linkTag) =>
+      shouldPromoteStylesheetPreloadTag(linkTag) ? promoteStylesheetPreloadTag(linkTag) : linkTag,
+  );
+
+  return {
+    gatedHtmlBuffer: gatedHtml === htmlString && !nextIncompleteLinkTagTail ? html : Buffer.from(gatedHtml),
```
[the remark this answers is not in the document]
Comment (justin808) [reply packages/react-on-rails-pro/src/injectRSCPayload.ts:367]: Fixed in 1d45f32cf. The streamed preload promotion path now treats incomplete UTF-8 tails like incomplete link-tag tails and waits for the next chunk before rewriting decoded HTML, so it never re-encodes a buffer ending mid-codepoint. Added a regression where an RSC CSS preload is promoted while `é` is split across HTML chunks, asserting the final stream preserves `café` and contains no U+FFFD replacement characters.

File: packages/react-on-rails-pro/src/injectRSCPayload.ts [hunk 9]
```diff
@@ -364,6 +711,14 @@ export default function injectRSCPayload(
...
                 }
               }
+              stylesheetTagsForRSCClientChunks(
+                flightData,
+                rscClientChunkStylesheetHrefsByChunkName,
+                emittedRSCClientStylesheetHrefs,
+              ).forEach((stylesheetTag) => {
+                rscClientStylesheetBuffers.push(Buffer.from(stylesheetTag));
+              });
+              resolveRSCClientStylesheetInferenceForStream();
```
[the remark this answers is not in the document]
Comment (justin808) [reply packages/react-on-rails-pro/src/injectRSCPayload.ts]: Fixed in 69251bdd4. The per-stream inferred stylesheet gate now remains pending through Flight chunks that do not produce any client stylesheet tags; it resolves only when a stylesheet is inferred, when the stream settles, or via the existing fail-open timeout. Added a regression where a stream emits server-only Flight data before a later client chunk import, and the final output now asserts the inferred stylesheet link precedes the Suspense reveal HTML.

**Comment by justin808** [comment]
Merge Ledger Block: formal review decision needed

Current head: `69251bdd4ae4efbf04e1457e128abf02eb994270`

Closeout evidence:
- GitHub checks are pass/skipped for the current head.
- GraphQL unresolved current-head review threads: 0.
- Strict merge ledger artifact: `/tmp/pr-batch-4031-4032-4009/pr-4047-merge-ledger-resume.json`.
- Ledger status: `complete_allowed: false` only because `pr.review_decision` is `UNKNOWN`.

Maintainer action requested: leave a formal GitHub review decision (approve, request changes, or comment with the intended review state). After that, rerun `script/pr-merge-ledger 4047 --repo shakacode/react_on_rails --changelog-classification changelog_present --finding-dispositions /tmp/pr-batch-4031-4032-4009/pr-4047-finding-dispositions.json --strict --pretty` before any merge/readiness claim.

**Comment by justin808** [comment]
Approved by maintainer, pending more adversarial AI reviews.

**Comment by justin808** [comment]
Merge Ledger Block: formal review decision still needed

Current head: `69251bdd4ae4efbf04e1457e128abf02eb994270`

Closeout evidence refreshed after adversarial review:
- GraphQL state: `mergeStateStatus=CLEAN`, unresolved review threads `0`, `reviewDecision=null`.
- GitHub checks are pass/skipped for the current head, including full CI and the newly requested benchmark suites.
- Benchmark run: https://github.com/shakacode/react_on_rails/actions/runs/27582027490 passed `Core benchmarks`, `Pro benchmarks (shard 1/2)`, `Pro benchmarks (shard 2/2)`, and `Pro Node Renderer benchmarks`.
- Local adversarial proof: PR tests fail against `origin/main` implementation and pass after restoring the PR implementation; focused Rspack/RSC Playwright FOUC spec passed locally (`6/6`).
- Strict merge ledger artifact: `/tmp/pr-batch-4031-4032-4009/pr-4047-merge-ledger-resume5.json`.
- Ledger status: `complete_allowed: false` only because `pr.review_decision` is `UNKNOWN`; P2 findings are marked fixed with evidence.

Maintainer action requested: submit a formal GitHub review decision for the current head. A regular PR comment such as “approved” still leaves GitHub `reviewDecision` null, so the strict ledger continues to block merge/readiness.

**Comment by justin808** [comment]
Maintainer Waiver And Immediate Merge Decision

Current head: `69251bdd4ae4efbf04e1457e128abf02eb994270`

Final live gate snapshot before merge:
- Release mode: accelerated-rc via #3823.
- Merge state: `CLEAN`.
- Unresolved review threads: `0`.
- Checks: all current-head checks are pass/skipped, including full CI and benchmark suites.
- Strict merge ledger: `/tmp/pr-batch-4031-4032-4009/pr-4047-merge-ledger-final.json`.
- Ledger status: `complete_allowed: false` only because GitHub `reviewDecision` is `UNKNOWN` / null.

Final waiver set:
- The formal non-author GitHub review object / `reviewDecision` gate is explicitly waived for this PR.
- The maintainer approval already provided in the Codex session is accepted as the merge authorization for this PR.
- Immediate merge is approved; do not wait for the default waiver-soak window.

Confidence note:
- Validated: focused Jest/Ruby proof failed against the old implementation and passed with this PR implementation; focused Rspack/RSC Playwright FOUC spec passed locally (`6/6`); GitHub checks and benchmark suites passed for the current head.
- Evidence: GitHub checks on this PR, benchmark run https://github.com/shakacode/react_on_rails/actions/runs/27582027490, local artifacts under `/tmp/pr-batch-4031-4032-4009/adversarial-4047/`.
- UNKNOWN: GitHub `reviewDecision` remains null by waiver.
- Residual risk: low; the change is release-sensitive, but the bug was reproduced without the fix and covered by focused unit/integration/E2E evidence plus benchmark CI.

**Comment by justin808** [comment]
## Manual verification: reproduced the bug and confirmed the fix ✅

Reproduced via the merged jest regression suites (`injectRSCPayload.test.ts`, `ClientSideRenderer.test.ts`) using the pre-fix-on-one-file tactic. Bug (#4031/#4032): streamed RSC client-chunk stylesheets were injected as `<link rel="preload" as="style">` hints (and not gated before reveal), so content revealed before its CSS applied → FOUC.

### Reproduction

Squash-merged at `af6525d8b`, so pre-fix = `af6525d8b~1`. Reverted only `src/injectRSCPayload.ts` and `src/ClientSideRenderer.ts`, then restored.

```
cd packages/react-on-rails-pro && pnpm exec jest tests/injectRSCPayload.test.ts tests/ClientSideRenderer.test.ts --runInBand
```

### Results

| Scenario | Streamed RSC CSS handling | Result |
| --- | --- | --- |
| Pre-fix source | stays `<link rel="preload" as="style">`; not promoted/gated before reveal HTML | **9 failed / 53** (BROKEN) |
| Post-fix restored | promoted to `<link rel="stylesheet" data-precedence="rsc-css">` and emitted **before** the reveal `<div>` | **53 passed / 53** (FIXED) |

```text
# PRE-FIX (af6525d8b~1)
✕ promotes streamed RSC client chunk stylesheet preloads to gate reveal
    Expected: <link rel="stylesheet" ... data-precedence="rsc-css"/>
    Received: <link rel="preload" as="style" .../>      ← still just a preload hint, CSS not applied before reveal
✕ injects inferred RSC client chunk stylesheets before streamed reveal HTML   (stylesheetIndex < revealHtmlIndex fails)
✕ waits for inferred RSC client chunk stylesheets before flushing reveal HTML
   ... (9 FOUC-gating cases total)
Tests: 9 failed, 44 passed, 53 total

# POST-FIX (restored, git status clean)
Tests: 53 passed, 53 total
```

The 9 failures are exactly the CSS-before-reveal / preload-promotion gating cases.

### Caveat

Mechanism-level against the real `injectRSCPayload` stream output and `ClientSideRenderer` in jsdom — it confirms the promoted `<link rel="stylesheet">` is emitted before the reveal HTML and preload attributes/split-chunk handling are preserved. I did **not** run the Playwright E2E (`rsc_fouc.spec.ts`) in a real browser to observe zero visual flash, since that needs a built Pro dummy app + browser. Note this area has had revert churn (#3587 → #3860); this verification reflects net current behavior at HEAD, which now gates reveal on the promoted stylesheet.
